### PR TITLE
Create `pad_to_max_dim_and_stack()` function in torch_utils

### PR DESCRIPTION
**Pull request by xrdaukar**
# Description

-- The function pads input tensors to maximum size in each dimension, then stacks them
-- This is similar to `pad_sequences()` except it can pad to more than one dimension, which is needed e.g., for `cross_attention_mask` which has two variable dimensions: `seq_len`, `num_images` for multi-image multimodal examples
-- The integration of this new function into data pipeline is coming in a separate PR.
-- Minor error logging improvements in padding utils

## Related issues

Towards OPE-355

## Before submitting

- [ ] This PR only changes documentation. (You can ignore the following checks in that case)
- [X] Did you read the [contributor guideline](https://github.com/oumi-ai/oumi/blob/main/CONTRIBUTING.md) Pull Request guidelines?
- [X] Did you link the issue(s) related to this PR in the section above?
- [X] Did you add / update tests where needed?

## Reviewers

At least one review from a member of `oumi-ai/oumi-staff` is required.

### CODE REVIEW

File: src/oumi/utils/torch_utils.py [hunk 1]
```diff
@@ -441,6 +448,155 @@ def pad_sequences(
...
+                if curr_size < max_size:
+                    target_view = target_view.narrow(dim_idx, start=0, length=curr_size)
+
+        assert target_view.shape == input_tensor.shape
+        target_view[...] = input_tensor
+
+    return result
+
+
+def pad_to_max_dim_and_stack(
```
Comment (taenin) [inline src/oumi/utils/torch_utils.py:541]: It's a bit confusing that you have a private method with the exact same name without the leading underscore. Consider changing `_pad_to_max_dim_and_stack` -> `_pad_to_max_dim_and_stack_impl`?

File: src/oumi/utils/torch_utils.py [hunk 2]
```diff
@@ -441,6 +448,155 @@ def pad_sequences(
...
+    tensors_list: list[T],
+    *,
+    padding_value: float = 0,
+    padding_side: Optional[str] = None,
```
Comment (taenin) [inline src/oumi/utils/torch_utils.py:545]: Would the typing `Literal["left", "right"] | None` make more sense here?

**Review by taenin — APPROVED** [review]

File: src/oumi/utils/torch_utils.py (hunk 1 above)
Comment (xrdaukar) [reply src/oumi/utils/torch_utils.py:541]: updated

File: src/oumi/utils/torch_utils.py (hunk 2 above)
Comment (xrdaukar) [reply src/oumi/utils/torch_utils.py:545]: I tried to do it but it leads to complications e.g., some callers store `padding_side` as  unvalidated `str`, which would make it difficult to use this function, so I decided to keep it as is

File: src/oumi/utils/torch_utils.py [hunk 3]
```diff
@@ -441,6 +448,155 @@ def pad_sequences(
...
+    return [
+        _DimMinMaxSizes(
+            dim_index=idx, min_size=min_dim_sizes[idx], max_size=max_dim_sizes[idx]
+        )
+        for idx in range(num_dims)
+    ]
+
+
+def _pad_to_max_dim_and_stack(
+    tensors_list: list[torch.Tensor],
```
[the remark this answers is not in the document]
Comment (xrdaukar) [reply src/oumi/utils/torch_utils.py:494]: Good point.  I think it should be a larger sweep over Oumi codebase if we decide to do that: many functions marked as `torch.Tensor`  may work inefficiently or break (?) if sparse tensors is passed. To my knowledge we don't have any `non-sparse` typing constraints anywhere , and don't really use sparse tensors.  Anyhow, maybe good idea to open a Linear task  to do something about it..
